% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/scUtilities.R
\name{SingleFeatureMatrix}
\alias{SingleFeatureMatrix}
\title{SingleFeatureMatrix}
\usage{
SingleFeatureMatrix(
  fragment,
  features,
  cells = NULL,
  process_n = 2000,
  sep = c("-", "-"),
  verbose = TRUE
)
}
\arguments{
\item{fragment}{A list of \code{\link{Fragment}} objects. Note that if
setting the \code{cells} parameter, the requested cells should be present in
the supplied \code{Fragment} objects. However, if the cells information in
the fragment object is not set (\code{Cells(fragments)} is \code{NULL}), then
the fragment object will still be searched.}

\item{features}{A GRanges object containing a set of genomic intervals.
These will form the rows of the matrix, with each entry recording the number
of unique reads falling in the genomic region for each cell.}

\item{cells}{Vector of cells to include. If NULL, include all cells found
in the fragments file}

\item{process_n}{Number of regions to load into memory at a time, per thread.
Processing more regions at once can be faster but uses more memory.}

\item{sep}{Vector of separators to use for genomic string. First element is
used to separate chromosome and coordinates, second separator is used to
separate start and end coordinates.}

\item{verbose}{Display messages}
}
\value{
SingleFeatureMatrix
}
\description{
Run FeatureMatrix on a single Fragment object
}
\keyword{internal}
